### PR TITLE
feat(azure-kv): added keychain plugin for azure keyvault

### LINKED ISSUE #971: feat(keychain-azure-kv): create azure key vault plugin
### Description
As a developer, I want to create a plugin for aws secret manager and perform tasks related to secret management like createSecret and retrieveSecret sort of functionalities.

For reference,
https://docs.microsoft.com/en-in/azure/key-vault/

### Acceptance Criteria
1. New package to be created as `@hyperledger/cactus-plugin-keychain-azure-kv`
2. Plugin must implement the central keychain plugin Typescript interface we have in the core-api package
3. The helper class to have methods to expose the credentials, port numbers, etc. that are required to configure/instantiate the `PluginKeychainAzureKv` class which is to be implemented as part of the package from point 1)

cc: @takeutak @sfuji822 @hartm @jonathan-m-hamilton @AzaharaC @jordigiam @kikoncuo @petermetz  @arnab-roy

### PR DESCRIPTION
**Pull request by jagpreetsinghsasan**
# Commit to be reviewed
------------------------------
feat(azure-kv): added keychain plugin for azure keyvault

        Primary Change
        ---
        1. Added new package cactus-plugin-keychain-azure-kv under packages/
        2. Added PluginKeychainAzureKvMock class to mock the functions of SecretClient under packages/cactus-plugin-keychain-azure-kv/src/test/typescript/mock/plugin-keychain-azure-kv-mock.ts

Resolves #971

Signed-off-by: Jagpreet Singh Sasan <jagpreet.singh.sasan@accenture.com>

### CODE REVIEW

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/generated/openapi/typescript-axios/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+wwwroot/*.js
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/generated/openapi/typescript-axios/.gitignore:1]: @jagpreetsinghsasan Please ignore this file in the openapi generator ignore file (and also remove the file itself from version control)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/generated/openapi/typescript-axios/.npmignore [hunk 2]
```diff
@@ -0,0 +1 @@
+# empty npmignore to ensure all required files (e.g., in the dist folder) are published by npm
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/generated/openapi/typescript-axios/.npmignore:1]: @jagpreetsinghsasan Please ignore this file in the openapi generator ignore file (and also remove the file itself from version control)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/generated/openapi/typescript-axios/git_push.sh [hunk 3]
```diff
@@ -0,0 +1,58 @@
+#!/bin/sh
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/generated/openapi/typescript-axios/git_push.sh:1]: @jagpreetsinghsasan Please ignore this file in the openapi generator ignore file (and also remove the file itself from version control)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
+import type { Server } from "http";
+import type { Server as SecureServer } from "https";
+
+import type { Express } from "express";
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts:4]: @jagpreetsinghsasan Thank you for adding `type` keyword, that's awesome. What this does is that you you can move the `express` dependency to `devDependencies` in the package.json file as well (and then run the configure script again to update the lock file accordingly)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  /**
+   * Dummy implementation that wires no web services on the host API server
+   * because there is no need. All the functionality is implemented somewhere
+   * else on a host that's accessible through the network to this object
+   * (because this class is a remote adapter not an actual plugin impl.).
+   *
+   * @param _expressApp
+   */
+  public async getOrCreateWebServices(): Promise<IWebServiceEndpoint[]> {
+    return [];
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts:86]: @jagpreetsinghsasan If not already, please create an issue to implement the web services (I may have already said this, please disregard if the issues are already in existence)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+  public getInstanceId(): string {
+    return this.instanceId;
+  }
+
+  public getPackageName(): string {
+    return `@hyperledger/cactus-plugin-keychain-azure-kv`;
+  }
+
+  public getAspect(): PluginAspect {
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts:153]: @jagpreetsinghsasan This method can be deleted IIRC. Once you rebase onto main the plugin aspect should cease to exist.

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+export interface IPluginKeychainAzureKvOptions extends ICactusPluginOptions {
+  logLevel?: LogLevelDesc;
+  keychainId: string;
+  instanceId: string;
+  azureEndpoint: string;
+  azureCredentialType: AzureCredentialType;
+  azureInMemoryCredentials?: IAzureInMemoryCredentials;
+}
+
+export class PluginKeychainAzureKv implements ICactusPlugin, IPluginWebService {
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:55]: @jagpreetsinghsasan Also implement the keychain interface from the core api package (as long as I remember correctly that we have one)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+  public static readonly CLASS_NAME = "PluginKeychainAzureKv";
+
+  readonly vaultUrl: string;
+  private readonly log: Logger;
+  private readonly instanceId: string;
+  private endpoints: IWebServiceEndpoint[] | undefined;
+  private azureKvClientMock: PluginKeychainAzureKvMock;
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:62]: @jagpreetsinghsasan Code in the ./src/main/... directory cannot import anything from the ./src/test/... directory. The trick is to rename this to `azureKvClient` (or `backend` if you want) and declare it's type as `SecretClient`.
Then add the same thing as a parameter to the constructor and for the tests pass in an instance of your mock class.

TLDR: There can be no trace of any mock types nor mock implementations in this class, it has to be abstracted out.
The way it's written now AFAICT if I deployed this to production today it would just be running the mock instead of using the real `SecretClient`.

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+   */
+  async get<T>(key: string): Promise<T> {
+    const keyVaultSecret: KeyVaultSecret = await this.azureKvClientMock.getSecret(
+      key,
+    );
+    if (keyVaultSecret) {
+      const result = keyVaultSecret.value;
+      return (result as unknown) as T;
+    } else {
+      return (false as unknown) as T;
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:218]: @jagpreetsinghsasan Throw if it doesn't exist. Then the endpoints can return HTTP 404 upstream and we can write tests asserting that they do (later)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+export class PluginKeychainAzureKv implements ICactusPlugin, IPluginWebService {
+  public static readonly CLASS_NAME = "PluginKeychainAzureKv";
+
+  readonly vaultUrl: string;
+  private readonly log: Logger;
+  private readonly instanceId: string;
+  private endpoints: IWebServiceEndpoint[] | undefined;
+  private azureKvClientMock: PluginKeychainAzureKvMock;
+  private readonly secrets: Map<string, string>;
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:63]: @jagpreetsinghsasan Do we need this?

File: packages/cactus-plugin-keychain-azure-kv/src/test/typescript/integration/plugin-keychain-azure-kv.test.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  PluginKeychainAzureKv,
+} from "../../../main/typescript/public-api";
+
+import { AzureCredentialType } from "../../../main/typescript/plugin-keychain-azure-kv";
+
+const logLevel: LogLevelDesc = "TRACE";
+
+test("get,set,has,delete alters state as expected for AzureCredentialType.InMemory", async (t: Test) => {
+  const options: IPluginKeychainAzureKvOptions = {
+    instanceId: uuidv4(),
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/test/typescript/integration/plugin-keychain-azure-kv.test.ts:18]: @jagpreetsinghsasan FYI this is where you can use the mock class and pass it in. I think you'll also need a little extra logic in the constructor of the plugin to examine if a `SecretClient` was passed in or not. If it was, then use that, otherwise proceed to examine if the credentials were passed in so that we could create our own `SecretClient` etc...

If the if conditions get too complicated/messy I can recommend the provider pattern where you'd define a provider function accepting credentials and then you'd have a real provider that instantiates `SecretClient` and another would just ignore the passed in credentials and return the mock.
This would definitely be the cleaner solution compared to what I outlined above in my first paragraph.

File: packages/cactus-plugin-keychain-azure-kv/src/test/typescript/unit/api-surface.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import test, { Test } from "tape";
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/test/typescript/unit/api-surface.ts:1]: @jagpreetsinghsasan Please add a `.test` suffix to the file name (a new-ish convention)

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 7 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:55]: Is this something new that we have added, because I dont see this one implemented for vault plugin as well.

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 8 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:62]: I have renamed this variable to azureKvClient, but in the plugin class constructor, we have this
image

What should I be doing about this? (Because for this initialization, we need to import the mock class as well)

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 9 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:218]: Actually, I initially had a throw statement when the key doesnt exist.
But then in the test case, it fails like this
image

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 10 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:63]: Yes, I am storing the secrets in this Map<string,string> structure

Comment (petermetz) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:62]: @jagpreetsinghsasan Then it should be in the mock class instead

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 9 above)
Comment (petermetz) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:213]: @jagpreetsinghsasan That's a good thing IMO, you can make the test work then by either asserting that the promise had been rejected (via `promise-tape`'s `rejects(...)` method) or just doing an old school try-catch and then asserting that the catch block was executed.

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 8 above)
Comment (petermetz) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts]: @jagpreetsinghsasan I'd extend the constructor parameters to have not just the azureCredentials, but also a `backend?: SecretClient` parameter. Then in the constructor I'd check if `options.backend` is truthy and if yes, I'd use that and ignore the azureCredentials.
This would allow the tests to instantiate the plugin with the mock and without the plugin needing to know anything about the mock or contain any implementation that's for the mock.

**Comment by petermetz** [comment]
@jagpreetsinghsasan See also this for a good example on mocking for unit tests: https://github.com/petermetz/cactus/tree/feat-object-store-plugin-ipfs

packages/cactus-plugin-object-store-ipfs/src/test/typescript/fixtures/mock/ipfs/ipfs-files-api-mock.ts
packages/cactus-plugin-object-store-ipfs/src/test/typescript/fixtures/mock/ipfs/ipfs-http-client-mock.ts
packages/cactus-plugin-object-store-ipfs/src/main/typescript/plugin-object-store-ipfs.ts

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts (hunk 5 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv-remote-adapter.ts:85]: Created issue #1009 for writing the README.md
Created issue #1010 for web services implementation

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 7 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:53]: Any update on this @petermetz ?

Comment (petermetz) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:53]: @jagpreetsinghsasan Oops, missed this comment earlier (helps if you tag me, then GitHub highlights it specifically)

Double checked it, it's been around since last year so not too new: 
packages/cactus-core-api/src/main/typescript/plugin/keychain/i-plugin-keychain.ts

File: packages/cactus-plugin-keychain-azure-kv/src/test/typescript/integration/plugin-keychain-azure-kv.test.ts [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    const valueAfterDelete = await plugin.get(key);
+    t.notok(valueAfterDelete, "valueAfterDelete falsy OK");
+  } catch (exp) {
+    t.throws(exp);
+  }
+
+  t.end();
+});
+
+test("get,set,has,delete alters state as expected for AzureCredentialType.LocalFile", async (t: Test) => {
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/test/typescript/integration/plugin-keychain-azure-kv.test.ts:65]: @jagpreetsinghsasan This test seems equivalent to the one above because when we specify the `backend` explicitly, the credential type is not taken into account at all if I read the constructor code right. Based on that (if correct) I recommend removing this second test case until we can actually implement it in the future (once we figure out how to emulate the azure backend for real)

File: packages/cactus-plugin-keychain-azure-kv/src/test/typescript/integration/plugin-keychain-azure-kv.test.ts [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  await plugin.delete(key);
+
+  const hasAfterDelete = await plugin.has(key);
+  t.false(hasAfterDelete, "hasAfterDelete === false OK");
+
+  try {
+    const valueAfterDelete = await plugin.get(key);
+    t.notok(valueAfterDelete, "valueAfterDelete falsy OK");
+  } catch (exp) {
+    t.throws(exp);
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/test/typescript/integration/plugin-keychain-azure-kv.test.ts:59]: @jagpreetsinghsasan How does this assertion work? I thought `throws` takes a function as the first parameter
https://github.com/substack/tape#tthrowsfn-expected-msg

File: packages/cactus-plugin-keychain-azure-kv/src/test/typescript/mock/plugin-keychain-azure-kv-mock.ts [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+      name: secretName,
+    };
+    return keyVaultSecret;
+  }
+
+  async beginDeleteSecret(
+    name: string,
+  ): Promise<PollerLike<PollOperationState<DeletedSecret>, DeletedSecret>> {
+    this.secrets.delete(name);
+    return (null as unknown) as PollerLike<
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/test/typescript/mock/plugin-keychain-azure-kv-mock.ts:67]: @jagpreetsinghsasan How difficult would it be to return something non-null that satisfies this `PollerLike` type?

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts (hunk 9 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:206]: Oh, where did it go? XD
I added the throws here and added the try catch block in test, now its gone from here, but it is there.
I am fixing this, along with the other review comment about the assertion

File: packages/cactus-plugin-keychain-azure-kv/src/test/typescript/mock/plugin-keychain-azure-kv-mock.ts (hunk 15 above)
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/test/typescript/mock/plugin-keychain-azure-kv-mock.ts:67]: I have changed it (will upload soon)
image

File: packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+      ) {
+        const azureCredentials = new UsernamePasswordCredential(
+          opts.azureInMemoryCredentials.azureTenantId,
+          opts.azureInMemoryCredentials.azureClientId,
+          opts.azureInMemoryCredentials.azureUsername,
+          opts.azureInMemoryCredentials.azurePassword,
+        );
+        this.azureKvClient = new SecretClient(this.vaultUrl, azureCredentials);
+      } else {
+        const azureCredentials = new DefaultAzureCredential();
```
[the remark this answers is not in the document]
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:120]: The credentials can be of 6 types.
EnvironmentCredential
ManagedIdentityCredential
SharedTokenCacheCredential
VisualStudioCredential
VisualStudioCodeCredential
AzureCliCredential
InteractiveBrowserCredential

And none of them access path from a file.
Only the EnvironmentCredential type accepts one path to the certificate and private key pair in PEM or PFX format, which can authenticate the App Registration (AZURE_CLIENT_CERTIFICATE_PATH)
but these vars need to be set as environment vars and cant just point to a file, like in AWS
AZURE_TENANT_ID 
AZURE_CLIENT_ID 
AZURE_CLIENT_SECRET
AZURE_CLIENT_CERTIFICATE_PATH 
AZURE_USERNAME 
AZURE_PASSWORD

[the remark this answers is not in the document]
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-keychain-azure-kv/src/main/typescript/plugin-keychain-azure-kv.ts:120]: No, I dont think so, we cannot pass in a file path for this one.
It can only take in the `DefaultAzureCredentialOptions` interface, which is 
image

**Comment by petermetz** [comment]
@jagpreetsinghsasan Should I give this one another go? (I see the code changes but you haven't yet clicked re-request review)

File: packages/cactus-plugin-keychain-azure-kv/src/main/json/openapi.json [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    "schemas": {
+    }
+  },
+  "paths": {
+    "/api/v1/plugins/@hyperledger/cactus-plugin-keychain-azure-kv/get-keychain-entry": {
+      "post": {
+        "x-hyperledger-cactus": {
+          "http": {
+            "path": "/api/v1/plugins/@hyperledger/cactus-plugin-keychain-azure-kv/get-keychain-entry",
+            "verbLowerCase": "get"
```
[the remark this answers is not in the document]
Comment (petermetz) [reply packages/cactus-plugin-keychain-azure-kv/src/main/json/openapi.json:22]: @jagpreetsinghsasan ?

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-keychain-azure-kv/src/main/json/openapi.json [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        },
+        "operationId": "getKeychainEntryV1",
+        "summary": "Retrieves the contents of a keychain entry from the backend.",
+        "parameters": [],
+        "requestBody": {
+          "$ref": "../../../../cactus-core-api/src/main/json/openapi.json#/components/requestBodies/keychain_get_entry_request_body"
+        },
+        "responses": {
+          "200": {
+            "$ref": "../../../../cactus-core-api/src/main/json/openapi.json#/components/responses/keychain_get_entry_200"
```
Comment (petermetz) [inline packages/cactus-plugin-keychain-azure-kv/src/main/json/openapi.json:33]: @elenaizaguirre @jagpreetsinghsasan Please coordinate your PRs merging in order to make sure that the openapi.json spec references are updated "at the right time" depending on who's PR gets merged first.

Context: Elena is working on a change that necessitates for us to use the http URLs instead of relative file paths.

**Review by petermetz — APPROVED** [review]
@elenaizaguirre @jagpreetsinghsasan Please coordinate your PRs merging in order to make sure that the openapi.json spec references are updated "at the right time" depending on who's PR gets merged first.

Context: Elena is working on a change that necessitates for us to use the http URLs instead of relative file paths.

**Comment by takeutak** [comment]
Would you wait my review since I'm asking about this PR on #1017?

**Review by takeutak — APPROVED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
@jagpreetsinghsasan  Please update to all Cactus package version occurrences to 0.6.0 from 0.5.0 (just issued the release) and then run the configure script to also update the lock files as well. Commit all of those changes and then we are ready to merge. Don't forget the squash the commits into a single one.

**Review by petermetz — APPROVED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
@jagpreetsinghsasan Almost forgot, please also 
1. migrate the tsconfig.json inside your package to match the new boilerplate (see any of the other packages for an example).
2. Add an entry to the root tsconfig.json file under the `references` array that points to the new package that is being added in the PR
3. Do this to all the pending PRs that are adding new packages

**Review by petermetz — APPROVED** [review]
@jagpreetsinghsasan LGTM, please rebase onto latest upstream/main to resolve the conflict(s) and then we are good to go.

**Comment by petermetz** [comment]
@jagpreetsinghsasan (same thing a second time again)  Please rebase onto latest upstream/main to resolve the conflict(s) and then we are good to go.
